### PR TITLE
Add parameter to destroy upstream_memory_provider in finalize()

### PR DESCRIPTION
**Pull request by ldorau**
### Description

Add parameter to destroy `upstream_memory_provider` in `finalize()`.

### Checklist
- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

### CODE REVIEW

**Comment by bratpiorka** [comment]
why do we need this parameter? If the user creates an upstream provider, why can't he destroy it?

**Comment by ldorau** [comment]
> why do we need this parameter? If the user creates an upstream provider, why can't he destroy it?

In order to prevent a memory leak in our tests using the `umfPoolTest` fixture (see: #808) the upstream provider has to be destroyed automatically.

**Comment by igchor** [comment]
> > why do we need this parameter? If the user creates an upstream provider, why can't he destroy it?
> 
> In order to prevent a memory leak in our tests using the `umfPoolTest` fixture (see: #808) the upstream provider has to be destroyed automatically.

If it's only for tests then you can prevent the leak by just destroying the provider in pool_unique_handle_t. You just need to modify this line:
https://github.com/oneapi-src/unified-memory-framework/blob/e802af834525825be60dd6581b8b472338536801/test/poolFixtures.hpp#L53 

To be something like this:

```
auto dtor = [provider](umf_memory_pool_handle_t pool) {
    umfPoolDestory(pool):
    umfMemoryProviderDestory(provider);
};
return umf::pool_unique_handle_t(hPool, std;:move(dtor));
```

**Comment by ldorau** [comment]
> > > why do we need this parameter? If the user creates an upstream provider, why can't he destroy it?
> > 
> > 
> > In order to prevent a memory leak in our tests using the `umfPoolTest` fixture (see: #808) the upstream provider has to be destroyed automatically.
> 
> If it's only for tests then you can prevent the leak by just destroying the provider in pool_unique_handle_t. You just need to modify this line:
> 
> https://github.com/oneapi-src/unified-memory-framework/blob/e802af834525825be60dd6581b8b472338536801/test/poolFixtures.hpp#L53
> 
> To be something like this:
> 
> ```
> auto dtor = [coarse_provider ](umf_memory_pool_handle_t pool) {
>     umfPoolDestory(pool):
>     umfMemoryProviderDestory(coarse_provider );
> };
> return umf::pool_unique_handle_t(hPool, std;:move(dtor));
> ```

Thanks @igchor ! I applied that in #808.

**Comment by ldorau** [comment]
> > why do we need this parameter? If the user creates an upstream provider, why can't he destroy it?
> 
> @igchor @bratpiorka @ldorau
> 
> But what if the user creates a memory pool with the `UMF_POOL_CREATE_FLAG_OWN_PROVIDER` flag? In that case, the user expects that the pool destructor will destroy the memory provider used by the pool. And in case the coarse provider is used, it is destroyed automatically, but the upstream memory provider is not.
> 
> Did I miss something?

@vinser52 You are right. It seems this parameter is needed when `UMF_POOL_CREATE_FLAG_OWN_PROVIDER` is used.
@bratpiorka @lplewa Do you agree?

**Comment by bratpiorka** [comment]
@ldorau agree + please add test for this (for both cases - when UMF_POOL_CREATE_FLAG_OWN_PROVIDER is set to true and false)

**Comment by ldorau** [comment]
> @ldorau agree + please add test for this (for both cases - when UMF_POOL_CREATE_FLAG_OWN_PROVIDER is set to true and false)

@bratpiorka  Done

**Review by bratpiorka — APPROVED** [review]

**Review by lukaszstolarczuk — COMMENTED** [review]
👍

File: src/provider/provider_coarse.c [hunk 1]
```diff
@@ -898,6 +901,13 @@ static umf_result_t coarse_memory_provider_initialize(void *params,
         return UMF_RESULT_ERROR_INVALID_ARGUMENT;
     }
 
+    if (coarse_params->destroy_upstream_memory_provider &&
+        !coarse_params->upstream_memory_provider) {
```
Comment (vinser52) [inline src/provider/provider_coarse.c:905]: Should the coarse provider always require the upstream provider to be specified?

What is the reason for not specifying the upstream provider? @bratpiorka what do you think?

P.S. I understand that it is out of scope for this particular PR.

Comment (ldorau) [reply src/provider/provider_coarse.c:905]: The coarse provider contains the fixed provider yet - in the case of the fixed provider the upstream provider is not required. The fixed provider will be moved to a separate provider and the upstream provider will be always required then.
